### PR TITLE
[torch/elastic] Introduce `DynamicRendezvousHandler` and `RendezvousBackend`.

### PR DESCRIPTION
**Pull request by cbalioglu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55637 [torch/elastic] Introduce `EtcdRendezvousBackend`.
* #55636 [torch/elastic] Introduce `C10dRendezvousBackend`.
* **#55635 [torch/elastic] Introduce `DynamicRendezvousHandler` and `RendezvousBackend`.**

This diff introduces the `DynamicRendezvousHandler` type as a stub implementation and its accompanying `RendezvousBackend` interface.

`DynamicRendezvousHandler` is intended to be a backend-agnostic type that will contain the core (bulk) logic of rendezvous handling. Any backend specific operation will be delegated to a concrete subclass of `RendezvousBackend` (e.g. `C10dRendezvousBackend` - see D27654492) that is passed as a constructor argument to `DynamicRendezvousHandler`.

Differential Revision: [D27654478](https://our.internmc.facebook.com/intern/diff/D27654478/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27654478/)!

### CODE REVIEW

File: torch/distributed/elastic/rendezvous/api.py [hunk 1]
```diff
@@ -43,93 +42,80 @@ class RendezvousHandler(abc.ABC):
...
+        `is_closed()` and `set_closed()` have semantics of eventual propagation
+        and should not be used for synchronization. The intention is that if at
+        least one node decides the job is finished, it will close the rendezvous
+        and other nodes will soon observe this and stop running as well.
         """
-        Checks whether rendezvous for current job has been closed,
-        which means all future attempts to re-rendezvous (within same job) will
-        fail.
-
-        .. note:: ``is_closed`` and ``set_closed`` have semantics of eventual
```
Comment (pritamdamania87) [inline torch/distributed/elastic/rendezvous/api.py:78]: Is there a reason this note was removed?

File: torch/distributed/elastic/rendezvous/default_rendezvous.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        """See base class."""
+        raise NotImplementedError()
+
+    def get_run_id(self) -> str:
+        """See base class."""
+        return self._run_id
+
+    def shutdown(self) -> bool:
+        """See base class."""
+        raise NotImplementedError()
```
Comment (pritamdamania87) [inline torch/distributed/elastic/rendezvous/default_rendezvous.py:252]: I might be missing something here, but the PR says the bulk of the rendezvous logic would be in this class. But its not implemented here and I don't see it in the other two PRs on top of it (C10dRendezvous and EtcRendezvous backends). Is there another PR in the works that adds this functionality?

Comment (cbalioglu) [reply torch/distributed/elastic/rendezvous/default_rendezvous.py:252]: Yes, I briefly mentioned it in the description as "as a stub implementation". It is actually the core piece of this whole task and will have its own PR by tomorrow.

File: torch/distributed/elastic/rendezvous/api.py (hunk 1 above)
Comment (cbalioglu) [reply torch/distributed/elastic/rendezvous/api.py:78]: The note is still there actually. I just removed the "note::" marking and included the description in the main description.

File: test/distributed/elastic/rendezvous/default_rendezvous_test.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+        self.assertEqual(timeout.join, timedelta(seconds=600))
+        self.assertEqual(timeout.last_call, timedelta(seconds=30))
+        self.assertEqual(timeout.close, timedelta(seconds=30))
+
+    def test_init_raises_error_if_timeout_is_not_positive(self) -> None:
+        join_timeouts = [timedelta(seconds=0), timedelta(seconds=-1)]
+
+        for join_timeout in join_timeouts:
+            with self.subTest(join_timeout=join_timeout):
```
Comment (H-Huang) [inline test/distributed/elastic/rendezvous/default_rendezvous_test.py:44]: neat! I haven't seen `subTest` before

**Review by H-Huang — APPROVED** [review]
Left some comments on internal diff, accepting here. 

Just to confirm, the newly added tests for torch/elastic have yet to be included as part of OSS CI right? They are only run internally?

File: torch/distributed/elastic/rendezvous/dynamic_rendezvous.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+def create_handler(
+    store: Store, backend: RendezvousBackend, params: RendezvousParameters
+) -> DynamicRendezvousHandler:
+    """Create a new :py:class:`DynamicRendezvousHandler` from the specified
+    parameters.
+
+    +-------------------+------------------------------------------------------+
+    | Parameter         | Description                                          |
+    +===================+======================================================+
+    | join_timeout      | The total time, in seconds, within which the         |
```
Comment (tierex) [inline torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:271]: Do you think maybe it makes sense to use a single timeout for join_timeout, last_call_timeout and close_timeout ?
I wonder when users would need to explicitly set each one separately.

Comment (cbalioglu) [reply torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:271]: I think in most of the cases (if at all) the user would be just adjusting `join_timeout` only. Combining these timeouts into one does not make much sense in my opinion as they are used for different purposes. `close_timeout` is used for `set_closed()` and `shutdown()` and should ideally be much shorter than `join_timeout`. `last_call_timeout` on the other hand is used for early completion of the rendezvous and is mostly independent of `join_timeout`. `join_timeout` more or less determines how long a call to `next_rendezvous()` should take.

**Pull request closed without merging**